r3bbase/pars/R3BMSOffsetFinder.cxx

```diff
@@ -0,0 +1,174 @@
+// R3BMSOffsetFinder: Standard Constructor --------------------------
+R3BMSOffsetFinder::R3BMSOffsetFinder(const TString& name, Int_t iVerbose)
+    : FairTask(name, iVerbose)
+    , fMSOffset(0.0)
+    , fMinStatistics(1)
+    , fMSOffsetPar(nullptr)
+    , fSamplerMapped(nullptr)
+    , fSamplerMSMapped(nullptr)
+    , fh_Offset_Finder(nullptr)
+{
+}
+
+// R3BMSOffsetFinder: Destructor ----------------------------------------
+R3BMSOffsetFinder::~R3BMSOffsetFinder() { LOG(info) << "R3BMSOffsetFinder: Delete instance"; }
+
+void R3BMSOffsetFinder::SetParContainers()
+{
+    // Parameter Container
+    FairRuntimeDb* rtdb = FairRuntimeDb::instance();
+    if (rtdb == nullptr)
+    {
+        LOG(error) << "FairRuntimeDb not opened!";
+    }
+}
+
+// -----   Public method Init   --------------------------------------------
+InitStatus R3BMSOffsetFinder::Init()
+{
+
+    LOG(info) << "R3BMSOffsetFinder: Init";
+
+    FairRootManager* rootManager = FairRootManager::Instance();
+    if (rootManager == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMapped"));
+    if (fSamplerMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMSMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMSMapped"));
+    if (fSamplerMSMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    FairRuntimeDb* rtdb = FairRuntimeDb::instance();
+    if (rtdb == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fMSOffsetPar = dynamic_cast<R3BMSOffsetPar*>(rtdb->getContainer("MSOffsetPar"));
+    if (fMSOffsetPar == nullptr)
+    {
+        LOG(error) << "R3BMSOffsetFinder::Init() Couldn't get handle on MSOffsetPar container";
+        return kFATAL;
+    }
+
+    fh_Offset_Finder = R3B::root_owned<TH1F>("MSOffsetFinderHisto", "MSOffsetFinderHisto", 2000.0, -1000.0, 1000.0);
+
+    return kSUCCESS;
+}
+
+// -----   Public method ReInit   --------------------------------------------
+InitStatus R3BMSOffsetFinder::ReInit()
+{
+    SetParContainers();
+    return kSUCCESS;
+}
+
+// -----   Public method Exec   --------------------------------------------
+void R3BMSOffsetFinder::Exec(Option_t* /*opt*/)
+{
+
+    const Int_t sampHits = fSamplerMapped->GetEntriesFast();
+    const Int_t sampmsHits = fSamplerMSMapped->GetEntriesFast();
+    if ((sampHits == 0) || (sampmsHits != 1))
+    {
+        return;
+    }
+    Double_t SAMPTime = 0;
+    Double_t SAMPMSTime = 0;
+    R3BSamplerMappedData* SAMPMapped = nullptr;
+    R3BSamplerMappedData* SAMPMSMapped = nullptr;
+
+    SAMPMSMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMSMapped->At(0));
+    SAMPMSTime = SAMPMSMapped->GetTime();
+
+    for (Int_t i = 0; i < sampHits; i++)
+    {
+        SAMPMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMapped->At(i));
+        SAMPTime = SAMPMapped->GetTime();
+
+        // Fill the histogram
+        fh_Offset_Finder->Fill(SAMPTime - SAMPMSTime);
+    }
```

Comment:
Could use range-for loop.
```cpp
for(auto* SAMPMapped : TRangeDynCast<R3BSamplerMappedData>(fSamplerMapped))
{
    fh_Offset_Finder->Fill(SAMPMapped->GetTime() - SAMPMSTime);
}
```

Reply:
That is neat. Someone at ROOT seems to like modern C++.